Revize Snakemake závislostí

vasekp opened the pull request:
- označení souborů nepotřebných pro opětovný překlad jako `temp` (po úspěšném překladu se automaticky smažou, až už nejsou potřeba)
- totéž pro soubory, jejichž generace nic nestojí a zabírají zbytečně místo (jen kopie z `tpcb`)
- `_single_XXX.pdf` nechat v `cache_dir()` (fixes #28)
- odstranění závislostí, které se nepoužívají
- zjednodušení logiky kolem `.idx`, `.ind` a `.idx.old`, mezi překlady stačí zachovat `.ind`
- obecný cleanup

vasekp committed:
Revize Snakemake závislostí

* označení souborů nepotřebných pro opětovný překlad jako temp
* totéž pro soubory, jejichž generace nic nestojí a zabírají zbytečně místo (jen kopie z tpcb)
* _single.pdf nechat v cache_dir() (fixes #28)
* odstranění závislostí, které se nepoužívají
* zjednodušení logiky kolem .idx, .ind a .idx.old
* obecný cleanup

karel-brinda commented:
Latexové soubory (šablona) by neměly být `temp`. Je to např. kvůli ladění (pokud chci pokusně měnit texový zdroják např. v TeXStudiu). Nebo kvůli tomu, aby si člověk mohl zazálohovat latexové zdrojáky celého zpěvníku.

Jinak pozor na jednu specifickou vlastnost Snakemaku (kvůli které se snažím temp moc nepoužívat pro mezičlánky). Pokud mám workflow `soubor1` => `temp(soubor2)` => `soubor3` a změní se `soubor1`, tak se `soubor2` znovu generovat nebude, čili i `soubor3` zůstane neaktualizovaný (alespoň pokud si to dobře pamatuju; to je taková nešťastná odlišnost od Make).

karel-brinda commented:
Jinak už se může vyhodit i 

``` python
import PyPDF2
```

vasekp commented:
Nevšiml jsem si, že tam zůstal. Ale i když ne teď, je to jedno, bude se hodit brzo – plánuji ho použít na to vytváření `singles` z jednoho celkového souboru.

vasekp commented:
Ad předchozí komentář: v téhle situaci, když je pro `soubor3` potřeba, by se `soubor2` vygenerovat znovu musel, protože by přestal existovat. Rozdíly oproti `make` určitě budou, ale já jsem o tom nepřemýšlel jako o "jiném `make`", zkoušel jsem to brát jako prostě nový workflow.

Se zazálohováním zdrojáků pro vlastní kopii bez potřeby Snakemake souhlasím. Upravím a věci pro překlad v cache nechám. Počítám, že tím se vyřeší i tohle.

vasekp committed:
Nechat všechny zdroje pro ruční překlad

vasekp commented:
Teď se jako dočasné soubory mažou jen `.log` (při úspěšném překladu netřeba), `.out` (prázdný), `.pdf` v cache (ten zkopírovaný do hlavního adresáře je identický), `.idx` (jen mezistupeň pro `.ind`, při překladu regenerovaný), `.ind.old` a `_single_XXX.aux` (jednotlivé písničky `.aux` nepotřebují).

vasekp committed:
Možnost přeložit samostatné písně jako option (fixes #28)

options = [ "SINGLES" ]
vedle hlavního zpěvníku vyrobí samostatné písně jako PDF v cache_dir().

options = [ "SINGLESONLY" ]
vyrobí pouze ty, hlavní zpěvník přeskočí.

Protože se mění, co má být výstupní soubor, je potřeba změnit poslední řádku
Snakefile! Původní "cb_pdf()" znamená doslova "vyrob hlavní zpěvník", takže se
singles není kompatibilní. Aby tyto volby mohly mít platnost, je potřeba
přepsat "cb_pdf()" na "run()". Ve všech aktuálních Snakefile jsem tuto změnu
provedl pro konzistenci. Snakefile.test ukazuje příklad výroby singles.

Zpětná kompatibilita:
Když se ponechá "cb_pdf()", překlad bude fungovat přesně jako dříve. Stejně
tak, když se pouze změní "cb_pdf()" na "run()" bez změny options, nebo když se
jen změní options ale ponechá "cb_pdf()". Neaktualizované Snakefile budou
fungovat bez problémů.

vasekp commented:
Viz rozšířený popis posledního commitu.

karel-brinda commented:
Super! Jinak co se týče toho `run()`, tak to je přesně způsob, který používám např. u RNFtools (viz. např. http://rnftools.readthedocs.org/en/latest/tutorial/01_first_snakefile.html). Až `tpcb` bude samostatný balíček, tak se do zpěvníku bude vkládat akorát:

``` python
import tpcb

...

include: tpcb.include()
rule: input: tpcb.input()
```
